### PR TITLE
feat(auth): add typings to request contexts & bodies

### LINKED ISSUE #351: Validate GNAP requests with OpenAPI
From https://github.com/interledger/rafiki/pull/343#discussion_r880691528:

> 🤔 Maybe this OpenAPI validation middleware (#339) could be a workspace in the monorepo: [https://github.com/interledger/rafiki/tree/bw-openapi/packages/backend/src/openapi](https://github.com/interledger/rafiki/tree/bw-openapi/packages/backend/src/openapi?rgh-link-date=2022-05-24T16%3A04%3A12Z) Then this could validate requests based on [https://github.com/interledger/open-payments/blob/master/auth-server-open-api-spec.yaml](https://github.com/interledger/open-payments/blob/master/auth-server-open-api-spec.yaml?rgh-link-date=2022-05-24T16%3A04%3A12Z)

See:
https://github.com/interledger/rafiki/pull/339/commits/4556d21258dc2c8c61d3b0f735b6a50d2f430b14

### PR DESCRIPTION
**Pull request by njlie**
## Changes proposed in this pull request

- Adds types to contexts for grant, interaction, and token endpoints.

## Context

Closes #351, or more precisely, closes the reason it was reopened.

## Checklist

- [ ] Related issues linked using `fixes #number`
- [ ] Tests added/updated
- [ ] Documentation added
- [ ] Make sure that all checks pass

### CODE REVIEW

File: packages/auth/src/grant/routes.ts [hunk 1]
```diff
@@ -203,9 +222,19 @@ async function getGrantDetails(
   ctx.body = { access: grant.access }
 }
 
+interface StartQuery {
+  clientName: string
+  clientUri: string
+}
```
Comment (wilsonianb) [inline packages/auth/src/grant/routes.ts:228]: `clientName` and `clientUri` should be added to the OpenAPI spec to be enforced:
https://github.com/interledger/rafiki/blob/18d2537a59ae9adb74d61b850132957b40f7caa6/packages/auth/src/openapi/id-provider.yaml#L41

File: packages/auth/src/grant/routes.ts [hunk 2]
```diff
@@ -309,9 +344,14 @@ async function handleGrantChoice(
   ctx.status = 202
 }
 
+interface FinishParams {
+  id: string
+  nonce: string
+}
```
Comment (wilsonianb) [inline packages/auth/src/grant/routes.ts:350]: Maybe `StartParams` and `FinishParams` could be consolidated into a single interface

File: packages/auth/src/app.ts [hunk 3]
```diff
@@ -36,6 +37,67 @@ export interface AppContextData extends DefaultContext {
...
+
+export type StartContext<QueryT, ParamsT> = InteractionContext<QueryT, ParamsT>
+export type GetContext<ParamsT> = InteractionContext<never, ParamsT>
+export type ChooseContext<ParamsT> = InteractionContext<never, ParamsT>
+export type FinishContext<ParamsT> = InteractionContext<never, ParamsT>
+
+export type IntrospectContext<BodyT> = TokenContext<BodyT>
+export type RevokeContext = ManagementContext
+export type RotateContext = ManagementContext
+
```
Comment (wilsonianb) [inline packages/auth/src/app.ts:100]: The calls to `createValidatorMiddleware` below should specify the type guarded context.
For example:
https://github.com/interledger/rafiki/blob/e067823ff1e0a1f9cb13b4b14587c52547db70b1/packages/auth/src/app.ts#L196-L206
should change to have
```
      createValidatorMiddleware<CreateContext>(openApi.authServerSpec, {
        path: '/',
        method: HttpMethod.POST
      })
```

That gets tricky without having the particular `BodyT` defined in this file.
One solution is to use:
https://github.com/interledger/rafiki/blob/e067823ff1e0a1f9cb13b4b14587c52547db70b1/packages/backend/src/app.ts#L112-L117
then have
```
      createValidatorMiddleware<Context<grantRoutes.create>>(openApi.authServerSpec, {
        path: '/',
        method: HttpMethod.POST
      })
```
in which case a lot of these contexts can be defined in their corresponding `routes.ts` file and be defined with the specific generic type (`BodyT`, etc)

File: packages/auth/src/accessToken/routes.ts [hunk 4]
```diff
@@ -6,6 +6,29 @@ import { AccessTokenService, Introspection } from './service'
...
+type ManagementContext = Omit<AppContext, 'request'> & {
+  request: ManagementRequest
+}
+
+interface IntrospectBody {
+  access_token: string
+}
+export type IntrospectContext = TokenContext<IntrospectBody>
+export type RevokeContext = ManagementContext
+export type RotateContext = ManagementContext
```
Comment (wilsonianb) [inline packages/auth/src/accessToken/routes.ts:30]:
```suggestion
type IntrospectContext = TokenContext<IntrospectBody>
type RevokeContext = ManagementContext
type RotateContext = ManagementContext
```
unless we do want to import in `app.ts`

File: packages/auth/src/app.ts [hunk 5]
```diff
@@ -193,109 +200,164 @@ export class App {
...
+      create: (ctx: AppContext) => Promise<void>
+      continue: (ctx: AppContext) => Promise<void>
+    } = grantRoutes
     this.publicRouter.post(
       '/',
-      createValidatorMiddleware(openApi.authServerSpec, {
-        path: '/',
-        method: HttpMethod.POST
-      }),
+      createValidatorMiddleware<ContextType<typeof grantCreate>>(
```
[the remark this answers is not in the document]
Comment (njlie) [reply packages/auth/src/app.ts:212]: It seems like the type of `ContextType` and the like don't seem to be compatible unless it's cast to `AppContext` like in the current implementation.
```
Type 'CreateContext' does not satisfy the constraint 'ExtendableContext & { state: DefaultState; } & DefaultContext & { body: unknown; response: { body: unknown; }; }'.
  Type 'CreateContext' is missing the following properties from type 'ExtendableContext': app, response, req, res, and 52 more.

(alias) type CreateContext = Omit<AppContext, "request"> & {
    request: GrantRequest<GrantRequestBody, ParsedUrlQuery>;
    clientKeyId: string;
}
```

[the remark this answers is not in the document]
Comment (wilsonianb) [reply packages/auth/src/app.ts:212]: :thinking: 
Maybe this `extends Koa.ParameterizedContext` is causing us more trouble than it's worth?
https://github.com/interledger/rafiki/blob/8054dc567267f7a329198c150b46a623d95a6f8b/packages/openapi/src/middleware.ts#L5

[the remark this answers is not in the document]
Comment (wilsonianb) [reply packages/auth/src/app.ts:212]: either that or we should be less cavalier with `Omit`ing parts of `AppContext`
https://github.com/interledger/rafiki/blob/8054dc567267f7a329198c150b46a623d95a6f8b/packages/backend/src/app.ts#L70-L73

[the remark this answers is not in the document]
Comment (wilsonianb) [reply packages/auth/src/app.ts:212]: I vote we remove the `extends Koa.ParameterizedContext` in `openapi`'s `createValidatorMiddleware` (assuming that makes using our desired contexts possible)

(I'm not sure if that will help with your previous session type issue though.)

[the remark this answers is not in the document]
Comment (njlie) [reply packages/auth/src/app.ts:212]: It seems like the general issue is that `Koa.ParameterizedContext`  doesn't seem to produce a type that's assignable to the base Koa context, or that at least the session middleware doesn't expect it as a possible type for a koa context.

All for removing `extends Koa.ParameterizedContext` if it works, though. It'll at least lift that restriction for the validator middleware.

[the remark this answers is not in the document]
Comment (wilsonianb) [reply packages/auth/src/app.ts:212]: Meanwhile, renovates over here saying: "Help me help you!"
- https://github.com/interledger/rafiki/pull/806

(I don't know if it'll help)

File: packages/auth/src/app.ts [hunk 6]
```diff
@@ -193,109 +206,137 @@ export class App {
     const openApi = await this.container.use('openApi')
     /* Back-channel GNAP Routes */
     // Grant Initiation
+    const {
+      create: grantCreate,
+      continue: grantContinue
+    }: {
+      create: (ctx: AppContext) => Promise<void>
+      continue: (ctx: AppContext) => Promise<void>
+    } = grantRoutes
```
Comment (wilsonianb) [inline packages/auth/src/app.ts:215]: I think these can be removed
```suggestion
```

**Pull request merged**